{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "46dcd076",
   "metadata": {},
   "source": [
    "# scRNAseq comprehensive analysis – 1.Scanpy QC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e062130",
   "metadata": {},
   "source": [
    "Import and define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5c3cca-c442-4ed7-9755-b2e88203fc13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "### warning\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "### Scientific/plotting stack\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import rc_context\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import seaborn as sns\n",
    "import colorcet as cc\n",
    "import itertools\n",
    "from pathlib import Path\n",
    "from scipy.stats import median_abs_deviation\n",
    "from scipy.sparse import csr_matrix, issparse\n",
    "### harmony\n",
    "import harmonypy as hm\n",
    "### Scanpy e AnnData\n",
    "import scanpy as sc\n",
    "import anndata2ri\n",
    "import anndata\n",
    "### decoupler\n",
    "import decoupler as dc\n",
    "\n",
    "### celltypist\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "### Graphics\n",
    "from adjustText import adjust_text\n",
    "import matplotlib.patheffects as pe\n",
    "\n",
    "### Logging\n",
    "import logging\n",
    "\n",
    "### RPy2 \n",
    "import rpy2.robjects as ro\n",
    "import rpy2.rinterface_lib.callbacks as rcb\n",
    "from rpy2.robjects import r\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "import hdf5plugin\n",
    "anndata2ri.activate()\n",
    "\n",
    "\n",
    "%load_ext rpy2.ipython\n",
    "# Remember to insert in <username> your pc username \n",
    "os.makedirs(\"Users/<username>/results/info_Plots/Harmony\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/info_Plots/Umap\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/Markers_Plots/Harmony\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/Markers_Plots/Umap\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/Markers_Plots/PreFiltering\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/Trajectory\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/Communications\", exist_ok=True)\n",
    "os.makedirs(\"Users/<username>/results/Proportions\", exist_ok=True)\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "\n",
    "def split_umap(adata, split_by, basis='X_umap', color_by=None, ncol=2, nrow=None, save_path=None, dpi=300, **kwargs):\n",
    "    categories = adata.obs[split_by].cat.categories\n",
    "    if nrow is None:\n",
    "        nrow = int(np.ceil(len(categories) / ncol))\n",
    "\n",
    "    fig, axs = plt.subplots(nrow, ncol, figsize=(5*ncol, 4*nrow))\n",
    "    axs = axs.flatten()\n",
    "    for i, cat in enumerate(categories):\n",
    "        ax = axs[i]\n",
    "        sc.pl.embedding(\n",
    "            adata,\n",
    "            basis=basis,\n",
    "            color=color_by if color_by else split_by,\n",
    "            groups=[cat],\n",
    "            ax=ax,\n",
    "            show=False,\n",
    "            title=cat,\n",
    "            **kwargs\n",
    "        )\n",
    "\n",
    "    for j in range(i+1, len(axs)):\n",
    "        fig.delaxes(axs[j])\n",
    "\n",
    "    plt.tight_layout()\n",
    "    if save_path:\n",
    "        fig.savefig(save_path, dpi=dpi, bbox_inches='tight')\n",
    "        print(f\"Saved figure to {save_path}\")\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "from adjustText import adjust_text\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def gen_mpl_labels(\n",
    "    adata,\n",
    "    groupby,\n",
    "    basis=\"X_umap\",           \n",
    "    exclude=(),\n",
    "    ax=None,\n",
    "    adjust_kwargs=None,\n",
    "    text_kwargs=None,\n",
    "    color_by_group=False,\n",
    "    show_labels=True,\n",
    "):\n",
    "    if adjust_kwargs is None:\n",
    "        adjust_kwargs = {\"text_from_points\": False}\n",
    "    if text_kwargs is None:\n",
    "        text_kwargs = {}\n",
    "\n",
    "    if basis not in adata.obsm_keys():\n",
    "        raise ValueError(f\"{basis} non trovato in adata.obsm!\")\n",
    "\n",
    "    coords = adata.obsm[basis]\n",
    "\n",
    "    medians = {}\n",
    "    for g in adata.obs[groupby].cat.categories:\n",
    "        if g in exclude:\n",
    "            continue\n",
    "        mask = adata.obs[groupby] == g\n",
    "        if mask.sum() == 0:\n",
    "            continue\n",
    "        group_coords = coords[mask.values]\n",
    "        medians[g] = np.median(group_coords, axis=0)\n",
    "\n",
    "    text_colors = {g: \"black\" for g in adata.obs[groupby].cat.categories}\n",
    "    if color_by_group and f\"{groupby}_colors\" in adata.uns:\n",
    "        for i, g in enumerate(adata.obs[groupby].cat.categories):\n",
    "            if g in exclude:\n",
    "                continue\n",
    "            text_colors[g] = adata.uns[f\"{groupby}_colors\"][i]\n",
    "\n",
    "    texts = []\n",
    "    if show_labels:\n",
    "        for g, (x, y) in medians.items():\n",
    "            txt = (ax if ax is not None else plt).text(\n",
    "                x, y, g, color=text_colors[g], **text_kwargs\n",
    "            )\n",
    "            texts.append(txt)\n",
    "\n",
    "        adjust_text(texts, **adjust_kwargs)\n",
    "\n",
    "    return texts\n",
    "\n",
    "\n",
    "def process(adata, resolution=0.4):    \n",
    "    if adata.X.toarray().max() > 100:\n",
    "        adata.X = adata.layers[\"raw_counts\"].copy()\n",
    "        sc.pp.normalize_total(adata, target_sum=10**4) \n",
    "    \n",
    "        sc.pp.log1p(adata)\n",
    "    else:\n",
    "        print(\"Data are already log trasformed\")\n",
    "    sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "    if adata.raw is None:\n",
    "        print(\"log1p data before filtering have been saved in .raw.X\")\n",
    "        adata.raw = adata\n",
    "    adata.layers[\"log1p\"] = adata.X.copy()\n",
    "   \n",
    "    adata = adata[:, adata.var['highly_variable']]\n",
    "    sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])\n",
    "    sc.pp.scale(adata, max_value=10)\n",
    "    sc.tl.pca(adata, svd_solver='arpack', random_state=1)\n",
    "    sc.pp.neighbors(adata, random_state=1)\n",
    "    sc.tl.leiden(adata, resolution=resolution, random_state=1, key_added= f\"leiden_{resolution}\")\n",
    "    sc.tl.umap(adata, random_state=1)\n",
    "    \n",
    "    return adata\n",
    "\n",
    "\n",
    "def do_harmony(adata_to_harmonize, resolution=0.4, batch_column='orig.ident'):\n",
    "    adata = adata_to_harmonize.copy()\n",
    "    adata.obs['batch'] = adata.obs[batch_column].astype(str)\n",
    "    adata.uns['orig_umap'] = adata.uns['umap']\n",
    "    adata.obsm['orig_X_umap'] = adata.obsm['X_umap']\n",
    "    ho = hm.run_harmony(\n",
    "        adata.obsm['X_pca'],     \n",
    "        adata.obs,               \n",
    "        'batch', random_state=42  \n",
    "    )\n",
    "    adata.obsm[f'X_harmony_{batch_column}'] = ho.Z_corr.T\n",
    "\n",
    "    sc.pp.neighbors(adata, random_state=42, use_rep=f'X_harmony_{batch_column}', key_added=f'harmony_nn_{batch_column}')\n",
    "    sc.tl.leiden(adata, random_state=42, resolution=resolution, neighbors_key=f'harmony_nn_{batch_column}', key_added=f\"harmony_leiden_{batch_column}_{resolution}\")\n",
    "    sc.tl.umap(adata, random_state=42, neighbors_key=f'harmony_nn_{batch_column}')\n",
    "\n",
    "    adata.uns[f'harmony_umap_{batch_column}'] = adata.uns['umap']\n",
    "    adata.obsm[f'X_harmony_umap_{batch_column}'] = adata.obsm['X_umap']\n",
    "\n",
    "    adata.uns['umap'] = adata.uns['orig_umap']\n",
    "    adata.obsm['X_umap'] = adata.obsm['orig_X_umap']\n",
    "    del adata.uns['orig_umap']\n",
    "    del adata.obsm['orig_X_umap']\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7813a8d7-9132-4c90-8393-d4fc7b62b575",
   "metadata": {},
   "source": [
    "## Set Environment "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7a2c95e",
   "metadata": {},
   "source": [
    "Change the directory you are working on into the directory containing the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a4ad675-306c-49df-8a06-bc96d32b5387",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"directory now: \", os.getcwd())\n",
    "os.chdir(\"Users/<username>/data/\")\n",
    "print(\"directory now: \", os.getcwd())\n",
    "print(\"directory files: \")\n",
    "for file in os.listdir():\n",
    "    print(f\"\\n{file}\")\n",
    "palette = sns.color_palette(\"husl\", 36)\n",
    "res = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75d7cd85",
   "metadata": {},
   "source": [
    "## Process and perform integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68b6546c",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"Users/<username>/data/\")\n",
    "if \"AdataNorm\" not in os.listdir():\n",
    "        adata =sc.read_h5ad(\n",
    "        \"AdataFiltered\")\n",
    "        adata.layers[\"raw_counts\"]=adata.X.copy()\n",
    "        adata_sel=process(adata, resolution= 1)\n",
    "        \n",
    "        if not adata_sel.obs_names.is_unique:\n",
    "            new_names = [\n",
    "            f\"{barcode}_{patient}\"\n",
    "            for barcode, patient in zip(adata_sel.obs_names, adata_sel.obs[\"orig.ident\"])\n",
    "            ]\n",
    "            adata_sel.obs_names = new_names\n",
    "        import openpyxl\n",
    "        file_path2 = \"ptsinfo.xlsx\"\n",
    "        df_clinici2 = pd.read_excel(file_path2)\n",
    "\n",
    "        df_clinici2 = df_clinici2[df_clinici2['orig.ident'].notna()]\n",
    "        df_clinici2 = df_clinici2.replace(r'^\\s*$', np.nan, regex=True)\n",
    "        df_clinici2 = df_clinici2.dropna(axis=0, how='all')\n",
    "        df_clinici2 = df_clinici2.dropna(thresh=round(len(df_clinici2) * 0.7), axis=1)\n",
    "        cols = ['orig.ident'] + [c for c in df_clinici2.columns if c != 'orig.ident']\n",
    "        df_clinici2 = df_clinici2[cols]\n",
    "        df_clinici2 = df_clinici2.astype('category')\n",
    "        df_clinici2 = df_clinici2.applymap(lambda x: x.strip() if isinstance(x, str) else x)\n",
    "        df_clinici2.columns = df_clinici2.columns.str.strip()\n",
    "        df_clinici2 = df_clinici2[df_clinici2[\"orig.ident\"].isin(adata_sel.obs[\"orig.ident\"].value_counts().keys())]\n",
    "        df_clinici2 = df_clinici2.set_index(\"orig.ident\")\n",
    "        adata_sel.obs[\"Condition\"] = adata_sel.obs[\"orig.ident\"].map(df_clinici2[\"Condition\"])\n",
    "        adata_sel = do_harmony(adata_sel,resolution=1, batch_column='orig.ident')\n",
    "        adata_sel = do_harmony(adata_sel,resolution=1, batch_column='Condition_site_T0')\n",
    "        adata_sel.write_h5ad(\n",
    "                \"AdataNorm\",\n",
    "                compression=hdf5plugin.FILTERS[\"zstd\"])\n",
    "else:\n",
    "    adata =sc.read_h5ad(\n",
    "        \"AdataFiltered\")\n",
    "    adata_sel=sc.read_h5ad(\n",
    "        \"AdataNorm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1a40cb0",
   "metadata": {},
   "source": [
    "# Automated Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24e6fd58",
   "metadata": {},
   "source": [
    "Create the custom model to perform automated annotation with celltypist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b463c9e",
   "metadata": {},
   "source": [
    "Here we:\n",
    "1. Create a copy of the processed Anndata without the highly variable genes mask in order to be compatible with celltypist\n",
    "2. Make the predictions on the New Anndata and transfer them into the previous one\n",
    "3. Overwrite the data as \"AdataNorm\" to save computational time in the future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9ecec22b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 106458 × 3195\n",
       "    obs: 'orig.ident', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_Erythroid', 'log1p_total_counts_Erythroid', 'pct_counts_Erythroid', 'outlier', 'mt_outlier', 'erythroid_outlier', 'n_genes_by_counts_outlier', 'scDblFinder_score', 'scDblFinder_class', 'leiden_1', 'Biopsy_site_T0', 'batch', 'harmony_leiden_orig.ident_1', 'harmony_leiden_Biopsy_site_T0_1', 'MBC_Atlas_pred', 'MBC_Atlas_score', 'Immhigh_pred', 'Immhigh_score', 'ImmLow_pred', 'ImmLow_score'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'harmony_leiden_Biopsy_site_T0_1', 'harmony_leiden_orig.ident_1', 'harmony_nn_Biopsy_site_T0', 'harmony_nn_orig.ident', 'harmony_umap_Biopsy_site_T0', 'harmony_umap_orig.ident', 'hvg', 'leiden_1', 'log1p', 'neighbors', 'over_clustering', 'pca', 'umap'\n",
       "    obsm: 'X_harmony_Biopsy_site_T0', 'X_harmony_orig.ident', 'X_harmony_umap_Biopsy_site_T0', 'X_harmony_umap_orig.ident', 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'log1p', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances', 'harmony_nn_Biopsy_site_T0_connectivities', 'harmony_nn_Biopsy_site_T0_distances', 'harmony_nn_orig.ident_connectivities', 'harmony_nn_orig.ident_distances'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "981921f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"MBC_Atlas_pred\" not in adata_sel.obs.columns:\n",
    "    adata_sel =do_harmony(adata_sel, resolution=res, batch_column='orig.ident') #Condition_site is not good enough\n",
    "    adata_sel.uns['umap'] = adata_sel.uns['harmony_umap_orig.ident'].copy()\n",
    "    adata_sel.uns['neighbors'] = adata_sel.uns['harmony_nn_orig.ident'].copy()\n",
    "    adata_sel.obsp['distances'] = adata_sel.obsp[ 'harmony_nn_orig.ident_distances'].copy()\n",
    "    adata_sel.obsp['connectivities'] = adata_sel.obsp['harmony_nn_orig.ident_connectivities'].copy()\n",
    "    adata_sel.obsm['X_umap'] = adata_sel.obsm['X_harmony_umap_orig.ident'].copy()\n",
    "    adata_sel.obsm['X_pca'] = adata_sel.obsm['X_harmony_orig.ident'].copy()\n",
    "    adata_sel.obs[\"leiden_1\"] = adata_sel.obs[\"harmony_leiden_orig.ident_1\"].copy()\n",
    "    \n",
    "    adata_complete=anndata.AnnData(X=adata_sel.raw.X\n",
    "                          ,obs=adata_sel.obs\n",
    "                          ,var=adata_sel.raw.var\n",
    "                          ,uns=adata_sel.uns\n",
    "                          ,obsm=adata_sel.obsm\n",
    "                          ,obsp=adata_sel.obsp)\n",
    "    adata_complete.X = adata_complete.X.toarray()\n",
    "    predictions = celltypist.annotate(adata_complete, model = 'Immune_All_High.pkl', majority_voting = True,  mode = 'prob match', p_thres = 0.6)\n",
    "    adata_pred = predictions.to_adata()\n",
    "    adata_sel.obs[\"Immhigh_pred\"] = adata_pred.obs.loc[\n",
    "        adata_sel.obs.index, \"majority_voting\"\n",
    "    ]\n",
    "    adata_sel.obs[\"Immhigh_score\"] = adata_pred.obs.loc[\n",
    "        adata_sel.obs.index, \"conf_score\"\n",
    "    ]   \n",
    "    del adata_pred\n",
    "\n",
    "    predictions = celltypist.annotate(adata_complete, model = 'Immune_All_Low.pkl', majority_voting = True,  mode = 'prob match', p_thres = 0.6)\n",
    "    adata_pred = predictions.to_adata()\n",
    "    adata_sel.obs[\"ImmLow_pred\"] = adata_pred.obs.loc[\n",
    "        adata_sel.obs.index, \"majority_voting\"\n",
    "    ]\n",
    "    adata_sel.obs[\"ImmLow_score\"] = adata_pred.obs.loc[\n",
    "        adata_sel.obs.index, \"conf_score\"\n",
    "    ]\n",
    "    del adata_pred\n",
    "    del adata_complete\n",
    "    #convert categories as strings\n",
    "    for col in adata_sel.obs.select_dtypes(['category']).columns:\n",
    "        adata_sel.obs[col] = adata_sel.obs[col].astype(str)\n",
    "    adata_sel.write_h5ad(\n",
    "            \"AdataNorm\",\n",
    "            compression=hdf5plugin.FILTERS[\"zstd\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cfc7e36",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sel.obs[\"median_score\"] = adata_sel.obs[[\"Immhigh_score\",\"ImmLow_score\"]].median(axis=1)\n",
    "adata_sel.obs[\"mean_score\"] = adata_sel.obs[[\"Immhigh_score\",\"ImmLow_score\"]].mean(axis=1)\n",
    "\n",
    "print( \"Min mean score:\" , round(adata_sel.obs[\"mean_score\"].min(),3))\n",
    "print( \"Max mean score:\" , round(adata_sel.obs[\"mean_score\"].max(),3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9902147",
   "metadata": {},
   "source": [
    "Plot the scores and the predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6660d12f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_sel, color=[\"Immhigh_score\",\"ImmLow_score\", \"mean_score\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89600141",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_sel, color=['n_genes_by_counts'\n",
    "                             , 'log1p_n_genes_by_counts'\n",
    "                             , 'log1p_total_counts'\n",
    "                             , 'pct_counts_mt'\n",
    "                             , 'pct_counts_ribo'\n",
    "                             , 'pct_counts_Erythroid'\n",
    "                             ]\n",
    "                             , ncols= 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dec53c8",
   "metadata": {},
   "source": [
    "# Commonly used Markers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0f4f649",
   "metadata": {},
   "source": [
    "Visualize commonly known Markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00793af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "res_dir = \"Users/<username>/results/\"\n",
    "fig = sc.pl.embedding(\n",
    "    adata_sel,\n",
    "    basis='X_umap',\n",
    "    color=[\n",
    "            'EPCAM','KRT5','KRT7','KRT8','KRT14','KRT18','KRT16','KRT17','CCN1', 'DST','TRPS1', 'CCSER1','ESR1','ERBB2', 'MKI67',\n",
    "            'PECAM1', 'PROX1', 'CD34',\n",
    "            'GNLY', 'NKG7', 'CD247',\n",
    "            'CD3D', 'CD3E',\n",
    "            'MS4A1',\n",
    "            'JCHAIN','CLEC4C','LILRA4',\n",
    "            'CD68',\n",
    "            'KIT',\n",
    "            'PDGFRA','PDGFRB', \n",
    "            'COL1A1', \n",
    "            'COL3A1', \n",
    "            'DCN',\n",
    "            'LUM',\n",
    "            'ACTA2','RGS5',\n",
    "            'ALB', 'APOB','FABP1', 'AFP'],\n",
    "    ncols=3,\n",
    "    vmin=0,\n",
    "    vmax=\"p99\",         \n",
    "    sort_order=False,   \n",
    "    cmap=\"viridis\",\n",
    "    return_fig=True,\n",
    "    wspace=0.5\n",
    ")\n",
    "\n",
    "\n",
    "fig.savefig(res_dir+'Markers_Plots/PreFiltering/KnownMarkersRAW.pdf',bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a9e9c8",
   "metadata": {},
   "source": [
    "# Manual Curation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cd08bf16",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = 2\n",
    "sc.tl.leiden(adata_sel, resolution=res, random_state=1, key_added= f\"leiden_{res}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6ee46ce",
   "metadata": {},
   "source": [
    "## Plot clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "289f3d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patheffects as pe\n",
    "plt.rcParams.update({'font.size': 10})\n",
    "with plt.rc_context({\"figure.figsize\": (12, 8), \"figure.dpi\": 150, \"figure.frameon\": False}):\n",
    "    fig = sc.pl.embedding(\n",
    "    adata_sel,\n",
    "    basis= \"X_umap\",\n",
    "    color=f\"leiden_{res}\",\n",
    "    show=False,\n",
    "    legend_loc='right margin',\n",
    "    frameon=False,\n",
    "    size=5,\n",
    "    add_outline=True,\n",
    "    return_fig=True\n",
    "    )\n",
    "    ax = fig.axes[0]\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=f\"leiden_{res}\",\n",
    "    basis= \"X_umap\",\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=4, foreground=\"black\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=f\"leiden_{res}\",\n",
    "    basis= \"X_umap\",\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(\"Users/<username>/results/Markers_Plots/PreFiltering/ClustersRAW.png\", dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "997e00b1",
   "metadata": {},
   "source": [
    "## Plot predictions and Biological conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22b57b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#predictions and/or biological conditions can hel with the classifications\n",
    "sc.pl.umap(adata_sel, color=[\"Condition\"\n",
    "                             , \"Immhigh_pred\"\n",
    "                             , \"ImmLow_pred\"]\n",
    "                             , ncols= 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4db50072",
   "metadata": {},
   "source": [
    "## Perform Differential gene expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb4ce998",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata_sel.copy()\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata, groupby=f\"leiden_{res}\"\n",
    "    , method=\"wilcoxon\"\n",
    "    , key_added=\"deg_clusters\"\n",
    ")\n",
    "\n",
    "\n",
    "sc.tl.filter_rank_genes_groups(\n",
    "    adata,\n",
    "    min_fold_change=1,\n",
    "    min_in_group_fraction=0.6,\n",
    "    max_out_group_fraction=0.4,\n",
    "    key=\"deg_clusters\",\n",
    "    key_added=\"deg_clusters_filtered\",\n",
    ")\n",
    "\n",
    "out = \"Users/<username>/results/Markers_Plots/cells_DGE.pdf\"\n",
    "os.makedirs(os.path.dirname(out), exist_ok=True)\n",
    "\n",
    "with plt.rc_context({'font.size': 18}):\n",
    "    dp = sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=f\"leiden_{res}\",\n",
    "    # standard_scale=\"var\",\n",
    "    vmin=0,\n",
    "    vmax=4,\n",
    "    n_genes=3,\n",
    "    key=\"deg_clusters_filtered\",\n",
    "    cmap='viridis',#'RdBu_r',  # \"viridis\",\n",
    "    #figsize=(60, 8),\n",
    "    min_logfoldchange=1,\n",
    "    #values_to_plot=\"logfoldchanges\",\n",
    "    #colorbar_title='log fold change',\n",
    "    colorbar_title='Mean expression',\n",
    "    show=True,  # Importante per ulteriori modifiche\n",
    "    return_fig=True,\n",
    "    swap_axes=True,\n",
    "    save = f\"_leiden{res}_DGE.pdf\")\n",
    "    dp.legends_width = 3        # <— allarga lo spazio delle legende\n",
    "    dp.savefig(out, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8a9d0c7",
   "metadata": {},
   "source": [
    "## Rename clusters accordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d229b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename every cluster according to Automated annotation, common markers and DGE\n",
    "# in this example we use 20 clusters but the number of clusters is experiment dependant\n",
    "plt.rcParams.update({'font.size': 6})\n",
    "adata_sel.obs[\"cell_type\"] = adata_sel.obs[f\"leiden_{res}\"].map(\n",
    "    {\n",
    "        \"0\": \"<insert cell type here>\",\n",
    "        \"1\": \"<insert cell type here>\",\n",
    "        \"2\": \"<insert cell type here>\",\n",
    "        \"3\": \"<insert cell type here>\",\n",
    "        \"4\": \"<insert cell type here>\",\n",
    "        \"5\": \"<insert cell type here>\",\n",
    "        \"6\": \"<insert cell type here>\",\n",
    "        \"7\": \"<insert cell type here>\",\n",
    "        \"8\": \"<insert cell type here>\",\n",
    "        \"9\": \"<insert cell type here>\",\n",
    "        \"10\": \"<insert cell type here>\",\n",
    "        \"11\": \"<insert cell type here>\",\n",
    "        \"12\": \"<insert cell type here>\",\n",
    "        \"13\": \"<insert cell type here>\",\n",
    "        \"14\": \"<insert cell type here>\",\n",
    "        \"15\": \"<insert cell type here>\",\n",
    "        \"16\": \"<insert cell type here>\",\n",
    "        \"17\": \"<insert cell type here>\",\n",
    "        \"18\": \"<insert cell type here>\",\n",
    "        \"19\": \"<insert cell type here>\",\n",
    "        \"20\": \"To delete\" # mark unknown / unwanted clusters with the To delete notation \n",
    "    }\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "adata_sel.obs[\"Condition\"] = (\n",
    "    adata_sel.obs[\"Condition\"]\n",
    "    .astype(str)   \n",
    "    .str.strip()   \n",
    "    .astype(\"category\")  \n",
    ")\n",
    "adata_sel.obs[\"cell_type\"] = adata_sel.obs[\"cell_type\"].astype(str)\n",
    "adata_sel.obs[\"Condition\"] = adata_sel.obs[\"Condition\"].astype(str)\n",
    "print( pd.value_counts(adata_sel.obs[\"cell_type\"]), adata_sel.n_obs )\n",
    "\n",
    "ax = sc.pl.embedding(adata_sel, color=\"cell_type\", legend_loc='right margin', frameon=False\n",
    "                    , size = 1, add_outline= True,title=\" \", basis=\"X_umap\")\n",
    "adata_sel = adata_sel[adata_sel.obs[\"cell_type\"] != \"To delete\"].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a75d38c",
   "metadata": {},
   "source": [
    "## Plot final annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d48f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patheffects as pe\n",
    "plt.rcParams.update({'font.size': 10})\n",
    "with plt.rc_context({\"figure.figsize\": (12, 8), \"figure.dpi\": 150, \"figure.frameon\": False}):\n",
    "    fig = sc.pl.embedding(\n",
    "    adata_sel,\n",
    "    basis= \"X_umap\",\n",
    "    color=\"cell_type\",\n",
    "    show=False,\n",
    "    legend_loc='right margin',\n",
    "    frameon=False,\n",
    "    size=5,\n",
    "    add_outline=True,\n",
    "    return_fig=True\n",
    "    )\n",
    "    ax = fig.axes[0]\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=\"cell_type\",\n",
    "    basis= \"X_umap\",\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=4, foreground=\"black\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=\"cell_type\",\n",
    "    basis= \"X_umap\",\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(\"Users/<username>/results/Markers_Plots/PreFiltering/CellsRAW.png\", dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b846814",
   "metadata": {},
   "source": [
    "## Redo processing but only on known cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c4c296d",
   "metadata": {},
   "source": [
    "In the annotation process one can filter out multiple cells. The intrinsic variation of the dataset can be influenced by that.\n",
    "\n",
    " For this reason we have to redo the normalization process and the UMAP construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "cad092bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "log1p data before filtering have been saved in .raw.X\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-03 14:29:14,540 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...\n",
      "2025-11-03 14:29:14 | [INFO] Computing initial centroids with sklearn.KMeans...\n",
      "2025-11-03 14:29:21,391 - harmonypy - INFO - sklearn.KMeans initialization complete.\n",
      "2025-11-03 14:29:21 | [INFO] sklearn.KMeans initialization complete.\n",
      "2025-11-03 14:29:21,567 - harmonypy - INFO - Iteration 1 of 10\n",
      "2025-11-03 14:29:21 | [INFO] Iteration 1 of 10\n",
      "2025-11-03 14:29:37,350 - harmonypy - INFO - Iteration 2 of 10\n",
      "2025-11-03 14:29:37 | [INFO] Iteration 2 of 10\n",
      "2025-11-03 14:29:53,052 - harmonypy - INFO - Iteration 3 of 10\n",
      "2025-11-03 14:29:53 | [INFO] Iteration 3 of 10\n",
      "2025-11-03 14:30:08,494 - harmonypy - INFO - Iteration 4 of 10\n",
      "2025-11-03 14:30:08 | [INFO] Iteration 4 of 10\n",
      "2025-11-03 14:30:23,632 - harmonypy - INFO - Iteration 5 of 10\n",
      "2025-11-03 14:30:23 | [INFO] Iteration 5 of 10\n",
      "2025-11-03 14:30:38,900 - harmonypy - INFO - Converged after 5 iterations\n",
      "2025-11-03 14:30:38 | [INFO] Converged after 5 iterations\n"
     ]
    }
   ],
   "source": [
    "#reload the AdataFiltered, subset it with adata_sel, use process functiomn\n",
    "adata =sc.read_h5ad(\n",
    "        \"AdataFiltered\")\n",
    "adata.layers[\"raw_counts\"]=adata.X.copy()\n",
    "\n",
    "if not adata.obs_names.is_unique:\n",
    "    adata.obs[\"orig.ident\"] = adata.obs[\"orig.ident\"].str.replace('_SF', '', regex=False)\n",
    "    # 1) Crea una nuova lista di nomi basata sull'indice e su \"orig.ident\"\n",
    "    new_names = [\n",
    "     f\"{barcode}_{patient}\"\n",
    "     for barcode, patient in zip(adata.obs_names, adata.obs[\"orig.ident\"])\n",
    "    ]\n",
    "\n",
    "# 2) Assegna la nuova lista di nomi come indice\n",
    "adata.obs_names = new_names\n",
    "adata = adata[adata_sel.obs_names].copy()\n",
    "adata.obs = adata_sel.obs\n",
    "adata_selOLD = adata_sel.copy()\n",
    "del adata_sel\n",
    "adata_sel =process(adata, resolution= 0.7)\n",
    "adata_sel = do_harmony(adata_sel,resolution=1, batch_column='orig.ident')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c99588b",
   "metadata": {},
   "source": [
    "## Plot the new UMAP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74acc4ed",
   "metadata": {},
   "source": [
    "Here we just plot the new obtained UMAP with the cell clusters, annotations, adn canonical markers.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80f0a723",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patheffects as pe\n",
    "plt.rcParams.update({'font.size': 10})\n",
    "with plt.rc_context({\"figure.figsize\": (12, 8), \"figure.dpi\": 150, \"figure.frameon\": False}):\n",
    "    fig = sc.pl.embedding(\n",
    "    adata_sel,\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    color=\"harmony_leiden_orig.ident_0.7\",\n",
    "    show=False,\n",
    "    legend_loc='right margin',\n",
    "    frameon=False,\n",
    "    size=5,\n",
    "    add_outline=True,\n",
    "    return_fig=True\n",
    "    )\n",
    "    ax = fig.axes[0]\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=\"harmony_leiden_orig.ident_0.7\",\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=4, foreground=\"black\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=\"harmony_leiden_orig.ident_0.7\",\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(\"Users/<username>/results/Markers_Plots/Umap/Clusters.png\", dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c531d93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dir = \"Users/<username>/results/\"\n",
    "fig = sc.pl.embedding(adata_sel, basis='X_harmony_umap_orig.ident',\n",
    "        color= ['EPCAM','KRT5','KRT7','KRT8','KRT14','KRT18','KRT16','KRT17','CCN1', 'DST','TRPS1', 'CCSER1','ESR1','ERBB2', 'MKI67',\n",
    "        'PECAM1', 'PROX1', 'CD34',\n",
    "        'GNLY', 'NKG7', 'CD247',\n",
    "        'CD3D', 'CD3E',\n",
    "        'MS4A1',\n",
    "        'JCHAIN','CLEC4C','LILRA4',\n",
    "        'CD68',\n",
    "        'KIT',\n",
    "        'PDGFRA','PDGFRB', \n",
    "        'COL1A1', \n",
    "        'COL3A1', \n",
    "        'DCN',\n",
    "        'LUM',\n",
    "        'ACTA2','RGS5',\n",
    "        'ALB', 'APOB','FABP1', 'AFP'],\n",
    "        ncols = 3,\n",
    "        vmin=0,\n",
    "        vmax=\"p99\",  # set vmax to the 99th percentile of the gene count instead of the maximum, to prevent outliers from making expression in other cells invisible. Note that this can cause problems for extremely lowly expressed genes.\n",
    "        sort_order=False,  # do not plot highest expression on top, to not get a biased view of the mean expression among cells\n",
    "        frameon=False,\n",
    "        cmap=\"viridis\", return_fig=True,wspace=0.5)  # or choose another color map e.g. from here: https://matplotlib.org/stable/tutorials/colors/colormaps.html )\n",
    "\n",
    "fig.savefig(res_dir+'Markers_Plots/Umap/KnownMarkers.png',bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b476f011",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patheffects as pe\n",
    "plt.rcParams.update({'font.size': 10})\n",
    "with plt.rc_context({\"figure.figsize\": (12, 8), \"figure.dpi\": 150, \"figure.frameon\": False}):\n",
    "    fig = sc.pl.embedding(\n",
    "    adata_sel,\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    color=\"cell_type\",\n",
    "    show=False,\n",
    "    legend_loc='right margin',\n",
    "    frameon=False,\n",
    "    size=5,\n",
    "    add_outline=True,\n",
    "    return_fig=True\n",
    "    )\n",
    "    ax = fig.axes[0]\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=\"cell_type\",\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=4, foreground=\"black\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "    gen_mpl_labels(\n",
    "    adata_sel,\n",
    "    groupby=\"cell_type\",\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    ax=ax,\n",
    "    text_kwargs=dict(fontsize=8, path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]),\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    color_by_group=True\n",
    "    )\n",
    "\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(\"Users/<username>/results/info_Plots/Harmony/CellsHRM.png\", dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbfdd2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import matplotlib.patheffects as pe\n",
    "plt.rcParams.update({'font.size': 10})\n",
    "with plt.rc_context({\"figure.figsize\": (12, 8), \"figure.dpi\": 150, \"figure.frameon\": False}):\n",
    "    fig = sc.pl.embedding(\n",
    "    adata_sel,\n",
    "    basis= 'X_harmony_umap_orig.ident',\n",
    "    color=\"Condition\",\n",
    "    show=False,\n",
    "    legend_loc='right margin',\n",
    "    frameon=False,\n",
    "    size=5,\n",
    "    add_outline=True,\n",
    "    return_fig=True\n",
    "    )\n",
    "\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(\"Users/<username>/results/info_Plots/Harmony/Condition_HRM.pdf\", dpi=300, bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "648e6041",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_umap(adata_sel,  basis='X_harmony_umap_orig.ident', split_by=\"Condition\", ncol= 4, na_in_legend = False, save_path=\"Users/<username>/results/info_Plots/Harmony/splitCondition_HRM.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d71e4138",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"Users/<username>/data/\")\n",
    "adata_sel.obs[\"Cells\"] = adata_sel.obs[\"cell_type\"]\n",
    "adata_sel.write_h5ad(\n",
    "            \"AdataFinal\",\n",
    "            compression=hdf5plugin.FILTERS[\"zstd\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d471a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"Users/<username>/data/\")\n",
    "adata_sel = sc.read_h5ad( \"AdataFinal\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "08a80f22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "Cells",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "b72f04b4-a76c-462f-9d30-745a6178ed1b",
       "rows": [
        [
         "MBC",
         "50820"
        ],
        [
         "Fibroblasts",
         "17942"
        ],
        [
         "VECs",
         "6707"
        ],
        [
         "T cells CD4+",
         "6147"
        ],
        [
         "Pericytes",
         "5830"
        ],
        [
         "T cells CD8+",
         "4462"
        ],
        [
         "Macrophages",
         "3882"
        ],
        [
         "Keratinocytes",
         "2798"
        ],
        [
         "Mast cells",
         "1110"
        ],
        [
         "B cells",
         "1048"
        ],
        [
         "Monocytes",
         "1031"
        ],
        [
         "NK cells",
         "772"
        ],
        [
         "LECs",
         "573"
        ],
        [
         "Hepatocytes",
         "383"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 14
       }
      },
      "text/plain": [
       "Cells\n",
       "MBC              50820\n",
       "Fibroblasts      17942\n",
       "VECs              6707\n",
       "T cells CD4+      6147\n",
       "Pericytes         5830\n",
       "T cells CD8+      4462\n",
       "Macrophages       3882\n",
       "Keratinocytes     2798\n",
       "Mast cells        1110\n",
       "B cells           1048\n",
       "Monocytes         1031\n",
       "NK cells           772\n",
       "LECs               573\n",
       "Hepatocytes        383\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_sel.obs.Cells.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20eac89b",
   "metadata": {},
   "source": [
    "## Final DE analysis to double check Cells identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "479e703a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata_sel.copy()\n",
    "sc.tl.rank_genes_groups(\n",
    "    adata, groupby=f\"Cells\"\n",
    "    , method=\"wilcoxon\"\n",
    "    , key_added=\"deg_clusters\"\n",
    ")\n",
    "\n",
    "\n",
    "sc.tl.filter_rank_genes_groups(\n",
    "    adata,\n",
    "    min_fold_change=1,\n",
    "    min_in_group_fraction=0.6,\n",
    "    max_out_group_fraction=0.4,\n",
    "    key=\"deg_clusters\",\n",
    "    key_added=\"deg_clusters_filtered\",\n",
    ")\n",
    "\n",
    "out = \"Users/<username>/results/Markers_Plots/cells_DGE.pdf\"\n",
    "os.makedirs(os.path.dirname(out), exist_ok=True)\n",
    "\n",
    "with plt.rc_context({'font.size': 18}):\n",
    "    dp = sc.pl.rank_genes_groups_dotplot(\n",
    "    adata,\n",
    "    groupby=f\"Cells\",\n",
    "    # standard_scale=\"var\",\n",
    "    vmin=0,\n",
    "    vmax=4,\n",
    "    n_genes=3,\n",
    "    key=\"deg_clusters_filtered\",\n",
    "    cmap='viridis',#'RdBu_r',  # \"viridis\",\n",
    "    #figsize=(60, 8),\n",
    "    min_logfoldchange=1,\n",
    "    #values_to_plot=\"logfoldchanges\",\n",
    "    #colorbar_title='log fold change',\n",
    "    colorbar_title='Mean expression',\n",
    "    show=True,  \n",
    "    return_fig=True,\n",
    "    swap_axes=True,\n",
    "    save = \"_Cells_DGE.pdf\")\n",
    "    dp.legends_width = 3        \n",
    "    dp.savefig(out, bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scvitest",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
